examples/nlp/cai/generate_sl_cai_dataset.py

```diff
@@ -0,0 +1,484 @@
+import argparse
+import json
+import os
+import os.path
+import random
+from typing import Optional
+
+import numpy as np
+import requests
+import tqdm
+from utils import remove_long_dialogs
+
+
+class MistralInstructChatTemplate:
+    bos_token = "<s>"
+    eos_token = "</s>"
+    user_prompt_start = "[INST]"
+    user_prompt_end = "[/INST]"
+    prompt_template = """{USER_PROMPT_START} {prompt} {USER_PROMPT_END}"""
+
+    @staticmethod
+    def apply_prompt_template(prompt):
+        return MistralInstructChatTemplate.prompt_template.format(
+            USER_PROMPT_START=MistralInstructChatTemplate.user_prompt_start,
+            prompt=prompt,
+            USER_PROMPT_END=MistralInstructChatTemplate.user_prompt_end,
+        )
+
+    @staticmethod
+    def extract_response(prompt: str):
+        response = prompt.rsplit(MistralInstructChatTemplate.user_prompt_end, 1)[-1]
+        response = response.strip().removesuffix(MistralInstructChatTemplate.eos_token)
+        return response
+
+
+def generate_chat_prompt(sample: dict):
+    dataset_name = "sl_cai_samples"
+    assert isinstance(sample, dict)
+    assert all(key in sample for key in ["initial_prompt", "revision_response"])
+    assert dataset_name is not None and dataset_name != ""
+
+    def _message_formatter(role: str, message: str):
+        assert role in ["User", "Assistant"]
+
+        return {"from": f"{role}", "value": f"{message}", "label": None}
+
+    def _conversation_formatter(messages: list, system_message: str = ""):
+        assert isinstance(dataset_name, str)
+        assert isinstance(messages, list) and len(messages) > 0
+
+        return {"system": f"{system_message}", "conversations": messages, "mask": "User", "dataset": f"{dataset_name}"}
+
+    prompt = _message_formatter(role="User", message=sample["initial_prompt"])
+    revision = _message_formatter(role="Assistant", message=sample["revision_response"])
+
+    return _conversation_formatter(messages=[prompt, revision])
+
+
+def model_remote_inference(
+    prompt,
+    port_num: int,
+    temperature: Optional[float] = 1.0,
+    greedy: Optional[bool] = True,
+    tokens_to_generate: Optional[int] = 1024,
+):
+    assert port_num >= 0
+    assert prompt is not None and isinstance(prompt, (str, list))
+    if not isinstance(prompt, list):
+        prompt = [prompt]
+    assert all(isinstance(p, str) for p in prompt)
+
+    headers = {"Content-Type": "application/json"}
+
+    def request_data(request):
+        resp = requests.put("http://localhost:{}/generate".format(port_num), data=json.dumps(request), headers=headers)
+        resp_json = resp.json()
+        resp_sentences = resp_json["sentences"]
+        return resp_sentences
+
+    data = {
+        "sentences": prompt,
+        "add_BOS": True,
+        "top_k": 1,
+        "top_p": 0.9,
+        "all_probs": False,
+        "repetition_penalty": 1.2,
+        "min_tokens_to_generate": 1,
```

Comment:
WDYT about parametrizing this?

Reply:
Thank you for your suggestion.
I've implemented the add_argument_group approach to parameterize the JSON object as you recommended.
i will do the same for 'examples/nlp/cai/generate_rl_cai_dataset.py'